libcontainer/container_linux.go

```diff
@@ -371,6 +371,10 @@ func (c *Container) Signal(s os.Signal) error {
 	// To avoid a PID reuse attack, don't kill non-running container.
 	switch status {
 	case Running, Created, Paused:
+	case Stopped:
+		if c.config.Namespaces.IsPrivate(configs.NEWPID) || !c.cgroupManager.Exists() {
```

Comment:
The problem here is we now have two different places in which we check `IsPrivate`. Also note that these two checks are slightly different (one also checks for cgroup existence, the other checks if `s` is SIGKILL. This is probably wrong.

Reply:
Another (minor) problem is with commit 9583b3d1c297021109 is it now calls Thaw even when we use signalAllProcesses, which is unnecessary.

Reply:
> the other checks if `s` is SIGKILL. This is probably wrong.

This is because you we remove ’-a’ option, so we also need to check SIGKILL here.

Reply:
What I meant it, there should be only one single check, deciding if we should use the usual logic (kill init) or the special case logic (kill all processes).

Something like 
```go
        if s == unix.SIGKILL && !c.config.Namespaces.IsPrivate(configs.NEWPID) {
                if err := signalAllProcesses(c.cgroupManager, unix.SIGKILL); err != nil {
                        return fmt.Errorf("unable to kill all processes: %w", err)
                }
                return nil
        }

// ... normal kill code goes here
```

plus
```diff
--- a/libcontainer/init_linux.go
+++ b/libcontainer/init_linux.go
@@ -633,6 +633,9 @@ func setupRlimits(limits []configs.Rlimit, pid int) error {
 // signalAllProcesses freezes then iterates over all the processes inside the
 // manager's cgroups sending the signal s to them.
 func signalAllProcesses(m cgroups.Manager, s unix.Signal) error {
+       if !m.Exists() {
+               return ErrNotRunning
+       }
        // Use cgroup.kill, if available.
        if s == unix.SIGKILL {
                if p := m.Path(""); p != "" { // Either cgroup v2 or hybrid.
```

Reply:
@kolyshkin I think maybe we should find back `runc kill -a`?
It is reasonable to remove `-a` with the signal `KILL`, but `runc kill` can pass other signals besides `KILL`, if we remove `-a` for runc-kill, it will have no way to send a signal to all the processes in the container.